components/script/canvas_state.rs

```diff
@@ -60,6 +60,46 @@ use crate::dom::paintworkletglobalscope::PaintWorkletGlobalScope;
 use crate::dom::textmetrics::TextMetrics;
 use crate::script_runtime::CanGc;
 
+// Firefox limits width/height to 32767 pixels and Chromium to 65535 pixels,
+// but slows down dramatically before it reaches that limit.
+// We limit by area instead, giving us larger maximum dimensions,
+// in exchange for a smaller maximum canvas size.
+const MAX_CANVAS_AREA: u64 = 32768 * 8192;
+// Max width/height to 65535 in CSS pixels.
+const MAX_CANVAS_SIZE: u64 = 65535;
+
+/// Information about the canvas draw target size and state.
+#[derive(Clone, Debug, JSTraceable, MallocSizeOf)]
+struct DrawTarget {
+    #[no_trace]
+    size: Size2D<u64>,
+    paintable: bool,
+}
+
+impl DrawTarget {
+    fn new(size: Size2D<u64>) -> DrawTarget {
+        DrawTarget {
+            size,
+            paintable: Self::is_supported_canvas_size(size),
+        }
+    }
+
+    fn paintable_size(&self) -> Size2D<u64> {
+        if self.paintable {
+            self.size
+        } else {
+            Size2D::zero()
+        }
+    }
```

Comment:
So we will use paintable size for toDataUrl,toBlobUrl and similar (hence in edge cases we will error), but when rendering we will simply return empty htmlcanvassource (transparent black image as handled by WR/Layout), right? Can we document this somewhere in code.